monix-execution/shared/src/main/scala/monix/execution/misc/AsyncSemaphore.scala

```diff
@@ -0,0 +1,197 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.execution.misc
+
+import monix.execution.{Cancelable, CancelableFuture}
+import monix.execution.atomic.AtomicAny
+import monix.execution.atomic.PaddingStrategy.LeftRight128
+
+import scala.annotation.tailrec
+import scala.collection.immutable.Queue
+import scala.concurrent.{ExecutionContext, Future, Promise}
+
+/** The `AsyncSemaphore` is an asynchronous semaphore implementation that
+  * limits the parallelism on `Future` execution.
+  *
+  * The following example instantiates a semaphore with a
+  * maximum parallelism of 10:
+  *
+  * {{{
+  *   val semaphore = AsyncSemaphore(maxParallelism = 10)
+  *
+  *   def makeRequest(r: HttpRequest): Future[HttpResponse] = ???
+  *
+  *   // For such a task no more than 10 requests
+  *   // are allowed to be executed in parallel.
+  *   val future = semaphore.greenLight(() => makeRequest(???))
+  * }}}
+  */
+final class AsyncSemaphore private (maxParallelism: Int)
+  extends Serializable {
+
+  import AsyncSemaphore.State
+  require(maxParallelism > 0, "parallelism > 0")
+
+  private[this] val stateRef =
+    AtomicAny.withPadding(AsyncSemaphore.initialState, LeftRight128)
+
+  /** Returns the number of active tasks that are holding on
+    * to the available permits.
+    */
+  def activeCount: Int =
+    stateRef.get.activeCount
+
+  /** Returns a new future, ensuring that the given source
+    * acquires an available permit from the semaphore before
+    * it is executed.
+    *
+    * The returned future also takes care of resource handling,
+    * releasing its permit after being complete.
+    *
+    * @param f is a function returning the `Future` instance we
+    *        want to evaluate after we get the permit from the
+    *        semaphore
+    */
+  def greenLight[A](f: () => Future[A])(implicit ec: ExecutionContext): Future[A] =
+    acquire().flatMap { _ =>
+      val result = f()
+      result.onComplete(_ => release())
+      result
+    }
+
+  /** Triggers a permit acquisition, returning a future that
+    * will complete when a permit gets acquired.
+    */
+  @tailrec def acquire(): CancelableFuture[Unit] = {
+    stateRef.get match {
+      case current @ State(activeCount, _, _) =>
+        if (activeCount < maxParallelism) {
+          val update = current.activateOne()
+
+          if (!stateRef.compareAndSet(current, update))
+            acquire() // retry
+          else
+            AsyncSemaphore.availablePermit
+        }
+        else {
+          val p = Promise[Unit]()
+          val update = current.addPromise(p)
```

Comment:
@ms-ati that current state is immutable and this is how you work with compareAndSet loops - you read the current value and try to update it. And no, it only gets added one, because if the CAS fails, we retry from scratch.